src/scripts/archive_dataset/archive_dataset.py

```diff
@@ -0,0 +1,141 @@
+and `--dataset-version=v2` will archive both `dataset_metadata_v2` and
+`dataset_metadata_files_v2`.
+
+Archive updates are tracked at the root dataset level.
+For example, if a struct `folders` is added to the `metadata_v2`
+JSON dataset in that dataset's second compatible schema update, then
+the next time a compatible schema update occurs the third archive
+operation will also occur. Although this is the first time that the
+`metadata_folders_v2` dataset will be archived, its archive will be
+named `metadata_folders_v2_3`.
+"""
+import argparse
+import shlex
+import subprocess
+import boto3
+
+
+def read_args():
+    parser = argparse.ArgumentParser(
+            description=("Archive a parquet dataset for each study of an app "
+                         "by copying it to the standard archive location: "
+                         "s3://{--bucket}/{--app}/{study}/parquet/archive/"
+                         "{--dataset}_{--dataset-version}_{update-number}/"))
+    parser.add_argument("--bucket",
+                        help="The name of the S3 bucket.")
+    parser.add_argument("--app",
+                        help="The app identifier.")
+    parser.add_argument("--dataset",
+                        help="The dataset name.")
+    parser.add_argument("--dataset-version",
+                        help="The dataset name.")
+    parser.add_argument("--profile",
+                        help="The AWS profile to use.")
+    args = parser.parse_args()
+    return args
+
+
+def get_source_and_dest_prefix(s3_client, bucket, app, dataset, dataset_version):
+    source_and_dest = dict()
+    study_prefix_obj = s3_client.list_objects_v2(
+            Bucket=bucket,
+            Prefix=f"{app}/",
+            Delimiter="/")
+    study_prefixes = [cp["Prefix"] for cp in study_prefix_obj["CommonPrefixes"]]
+    for study_prefix in study_prefixes:
+        parquet_prefix = f"{study_prefix}parquet/"
+        parquet_dataset_prefix_obj = s3_client.list_objects_v2(
+                Bucket=bucket,
+                Prefix=parquet_prefix,
+                Delimiter="/")
+        descendent_dataset_prefixes = [
+                cp["Prefix"] for cp in parquet_dataset_prefix_obj["CommonPrefixes"]
+                if f"{dataset}_" in cp["Prefix"] and dataset_version in cp["Prefix"]]
+        latest_archive_dataset_update = get_archive_dataset_update_number(
+                s3_client=s3_client,
+                bucket=bucket,
+                study_prefix=study_prefix,
+                dataset=dataset,
+                dataset_version=dataset_version)
+        for descendent_dataset_prefix in descendent_dataset_prefixes:
+            source_path = f"s3://{bucket}/{descendent_dataset_prefix}"
+            descendent_dataset = "_".join(
+                    descendent_dataset_prefix
+                    .split("/")[-2]
+                    .split("_")[:-1])
+            dest_path = (f"s3://{bucket}/{study_prefix}parquet/"
```

Comment:
Is there a typo here? Should it be `}/parquet`?  Also I recommend using `os.path.join`

Reply:
No typo, since the prefixes are returned with a trailing `/`, but I will import the os module and join these so that the code can be more clear.